src/globus_sdk/exc/api.py

```diff
@@ -157,67 +212,258 @@ def _get_args(self) -> list[t.Any]:
+            isinstance(subdoc, dict) for subdoc in self._dict_data["errors"]
         ):
-            # log a warning only when there is more than one error in the list
-            # if an API sends back an error of the form
-            #   {"errors": [{"foo": "bar"}]}
-            # then the envelope doesn't matter and there's only one error to parse
-            if len(json_data["errors"]) != 1:
-                log.warning(
-                    "Doing JSON load of error response with multiple "
-                    "errors. Exception data will only include the "
-                    "first error, but there are really %d errors",
-                    len(json_data["errors"]),
-                )
-            # try to grab the first error in the list, but also check
-            # if it isn't a dict
-            json_data = json_data["errors"][0]
-        self._load_from_json(json_data)
-
-    def _load_from_json(self, data: dict[str, t.Any]) -> None:
-        # rewrite 'code' if present and correct type
-        if isinstance(data.get("code"), str):
-            self.code = data["code"]
-
-        for f in self.MESSAGE_FIELDS:
-            if isinstance(data.get(f), str):
-                log.debug("Loaded message from '%s' field", f)
-                self.message = data[f]
-                break
+            raw_errors = self._dict_data["errors"]
+        # if no 'errors' were found, or 'errors' is invalid, then
+        # 'errors' should be set to contain the root document
+        else:
+            raw_errors = [self._dict_data]
+        self.errors = [
+            ErrorSubdocument(e, message_fields=self.MESSAGE_FIELDS) for e in raw_errors
+        ]
+
+        # use 'code' if present and correct type
+        if isinstance(self._dict_data.get("code"), str):
+            self.code = t.cast(str, self._dict_data["code"])
+        # otherwise, pull 'code' from the sub-errors array
+        elif self.errors:
+            # in undefined parse cases, the code will be left as `"Error"` for
+            # a higher degree of backwards compatibility
+            maybe_code = self._extract_code_from_error_array(self.errors)
+            if maybe_code is not None:
+                self.code = maybe_code
+
+        # either there is an array of subdocument errors or there is not,
+        # in which case we load only from the root doc
+        self.messages = self._extract_messages_from_error_array(
+            self.errors
+            or [ErrorSubdocument(self._dict_data, message_fields=self.MESSAGE_FIELDS)]
+        )
+
+        return True
+
+    def _extract_messages_from_error_array(
+        self, errors: list[ErrorSubdocument]
+    ) -> list[str]:
+        """
+        Extract 'messages' from an array of errors (JSON:API or otherwise)
+
+        This is done by checking each error document for the desired message fields and
+        popping out the first such message for each document.
```

Comment:
This comment seems a bit out of date.

Root error document is no longer responsible for determining what constitutes a subdocument "message" right?

Reply:
Let me review this docstring. I bet it needs a bit of revamping.

It gets a little difficult to describe perfectly -- you'll see cases where we pass `self.MESSAGE_FIELDS` to `ErrorSubdocument` objects. So the error class can influence behavior.

Reply:
Looking at it alongside the code, I described both what's happening in that method ("ask each subdoc for its message"), and the higher level fact that this means we're looking at fields which were chosen by the parsing path and the error class.
Let me know if you like the change and we can resolve this.